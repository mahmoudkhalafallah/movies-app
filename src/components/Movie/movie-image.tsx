import React, { useState } from "react";
import styled from "styled-components/native";
import { IMAGE_URI_PREFIX } from "../../utils/Constants";

const imagePlaceholder = require('../../../assets/images/film-poster-placeholder.png')

const StyledImage = styled.Image`
width: ${(props) => props.horizontal? '100' : '150'};
height: ${(props) => props.horizontal? '100' : '200'};
margin-right: ${(props) => props.horizontal? '10' : '20'};
border-top-left-radius: 10;
border-top-right-radius: 10;
border-bottom-left-radius: 10;
border-bottom-right-radius: 10;
`

export const MovieImage: React.FC<{ horizontal: string, src?: string }> = ({ horizontal, src }) => {

    const [err, setErr] = useState(false)
    const source = (!err && src) ? { uri: `${IMAGE_URI_PREFIX}${src}` } : imagePlaceholder

    return <StyledImage
        source={source}
        horizontal={horizontal}
        onError={() => setErr(true)}
    />
}