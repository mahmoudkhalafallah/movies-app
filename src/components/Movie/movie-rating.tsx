import React from "react";
import SwipeableRating from 'react-native-swipeable-rating';
import { Variables } from "../../utils/Variables";
import styled from "styled-components/native";
import { StyledText } from "../Text";


const RatingContainer = styled.View`
flex-direction: row;
align-items: center;
`

const RatingValue = styled(StyledText)`
font-size: 10;
color: #bbb;
`

export const MovieRating: React.FC<{ rating: number, horizontal: boolean }> = ({ rating, horizontal }) => {

    const ratingPerFiveStars = (rating / 10) * 5
    return <RatingContainer>
        <SwipeableRating
        color={Variables.ratingStarColor}
        emptyColor={Variables.ratingStarColor}
        rating={ratingPerFiveStars}
        size={15}
        gap={0.1}
        xOffset={20}
        allowHalves={true}
        swipeable={false}
    />
        {(!horizontal) && <RatingValue>{rating}/10</RatingValue>}
    </RatingContainer>
}
