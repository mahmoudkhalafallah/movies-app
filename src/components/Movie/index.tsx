import React from "react";

import { MovieTitle } from "../Titles";
import { MovieImage } from "./movie-image";
import { MovieRating } from "./movie-rating";
import styled from "styled-components/native";
import { StyledText } from "../Text";
import { IMAGE_URI_PREFIX } from "../../utils/Constants";

const MovieContainer = styled.View`
flex-direction: ${({ isMyMovie }: any) => isMyMovie ? 'column' : 'row'};
justify-content: space-between;
align-items: ${({ isMyMovie }: any) => isMyMovie ? 'center' : 'flex-start'};
width: 100;
margin-top: 10;
margin-bottom: 10;
${({ isMyMovie, isFirst }: any) => (isMyMovie && isFirst) ? 'margin-left: 25' : 'margin-left: 10'}
${({ isMyMovie, isLast }: any) => (isMyMovie && isLast) ? 'margin-right: 20' : 'margin-right: 10'}
${({ isMyMovie }: any) => (!isMyMovie) && 'margin-left: 20'}
${({ isMyMovie, isLast }: any) => (!isMyMovie && isLast) && 'padding-bottom: 80'}
`

const MovieDetails = styled.View`
flex-direction: column;
`

const MovieOverview = styled(StyledText)`
font-size: 13;
font-family: 'Oswald-Light';
margin-top: 5;
`

const MovieDate = styled(StyledText)`
font-size: 10;
margin-top: 5;
`

export const Movie: React.FC<any> = ({ title, poster_path, vote_average, overview,
    release_date, isFirst, isLast, isMyMovie }: any) => {
    return <MovieContainer isFirst={isFirst} isLast={isLast} isMyMovie={isMyMovie}>
        <MovieImage src={isMyMovie ? poster_path : `${IMAGE_URI_PREFIX}${poster_path}`} isMyMovie={isMyMovie}></MovieImage>
        <MovieDetails style={{width: isMyMovie ? 100 : 150}}>
            <MovieTitle ellipsizeMode='tail' numberOfLines={1}>{title}</MovieTitle>
            {(vote_average != undefined) && <MovieRating rating={vote_average} showRatingValue={isMyMovie} />}
            {!isMyMovie && <>
                <MovieDate>{release_date}</MovieDate>
                <MovieOverview ellipsizeMode='tail' numberOfLines={6}>{overview}</MovieOverview>
            </>}
        </MovieDetails>
    </MovieContainer>
}