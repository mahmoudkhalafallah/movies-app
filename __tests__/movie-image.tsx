import 'react-native';
import React from 'react';
import { MovieImage } from '../src/components/Movie/movie-image';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('Movie Image renders correctly with working source and isMyMovie flag set to true', () => {
    const rating = renderer.create(<MovieImage isMyMovie={true} src='https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png' />).toJSON();
    expect(rating).toMatchSnapshot();
});

it('Movie Image renders correctly with working source', () => {
    const rating = renderer.create(<MovieImage src='https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png' />).toJSON();
    expect(rating).toMatchSnapshot();
});

it('Movie Image renders correctly with broken source', () => {
    const rating = renderer.create(<MovieImage src='https://broken-image-src.png' />).toJSON();
    expect(rating).toMatchSnapshot();
});

it('Movie Image renders correctly with no source', () => {
    const rating = renderer.create(<MovieImage />).toJSON();
    expect(rating).toMatchSnapshot();
});
