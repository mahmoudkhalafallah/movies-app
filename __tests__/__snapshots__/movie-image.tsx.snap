// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Image renders correctly with broken source 1`] = `
<Image
  isMyMovie={false}
  onError={[Function]}
  source={
    Object {
      "uri": "https://broken-image-src.png",
    }
  }
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "height": 200,
        "marginRight": 20,
        "width": 150,
      },
    ]
  }
/>
`;

exports[`Movie Image renders correctly with no source 1`] = `
<Image
  isMyMovie={false}
  onError={[Function]}
  source={
    Object {
      "testUri": "../../../assets/images/film-poster-placeholder.png",
    }
  }
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "height": 200,
        "marginRight": 20,
        "width": 150,
      },
    ]
  }
/>
`;

exports[`Movie Image renders correctly with working source 1`] = `
<Image
  isMyMovie={false}
  onError={[Function]}
  source={
    Object {
      "uri": "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png",
    }
  }
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "height": 200,
        "marginRight": 20,
        "width": 150,
      },
    ]
  }
/>
`;

exports[`Movie Image renders correctly with working source and isMyMovie flag set to true 1`] = `
<Image
  isMyMovie={true}
  onError={[Function]}
  source={
    Object {
      "uri": "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png",
    }
  }
  style={
    Array [
      Object {
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "height": 100,
        "marginRight": 10,
        "width": 100,
      },
    ]
  }
/>
`;
